src/Container/ApplicationFactory.php

```diff
@@ -165,9 +166,20 @@ private function injectRoutes(Application $app, ContainerInterface $container)
             $methods = (isset($spec['allowed_methods']) && is_array($spec['allowed_methods']))
                 ? $spec['allowed_methods']
                 : null;
-            $name    = isset($spec['name']) ? $spec['name'] : null;
+            $name = isset($spec['name']) ? $spec['name'] : null;
             $methods = (null === $methods) ? Route::HTTP_METHOD_ANY : $methods;
-            $route   = new Route($spec['path'], $spec['middleware'], $methods, $name);
+
+            if (!is_callable($spec['middleware']) && is_array($spec['middleware'])) {
+                $middlewarePipe = new MiddlewarePipe();
+
+                foreach ($spec['middleware'] as $middleware) {
+                    $middlewarePipe->pipe(is_callable($middleware) ? $middleware : $container->get($middleware));
+                }
+
+                $spec['middleware'] = $middlewarePipe;
+            }
```

Comment:
@michaelmoussa In reviewing this for merge... I think the code above is in the wrong place! Everywhere else, we delay fetching from the container *unless we're dispatching the middleware*. As such, I think this needs to be moved to `routeMiddleware()`.

However, for that to happen, a few other changes are necessary:

- `RouteResult::getMatchedMiddleware()` (and its accompanying property, `$matchedMiddleware`, and the `$middleware` argument for `fromRouteMatch()`) would need to adjust the typehint to `string|array|callable`.
- `Route::getMiddleware()` (and its accompanying property `$middleware`, and the `$midleware` argument to the constructor) would need to adjust the typehint to `string|array|callable`.
- Similarly, we also need to update `Application::route()` (this needs updating anyways with your change!)

The above bit that creates a middleware pipe would then move into `routeMiddleware()`, between the `is_callable()` and `is_string()` checks. We could either inline it as you've done here, or create a method that returns the `MiddlewarePipe` instance, and call that.

If you go that route, you could likely move the bits from `is_string()` up to `return $callable()` into a method that could be used by both methods, for marshaling the individual middlewares.

So, now the question: do you want to tackle those changes, or do you want me to do a PR against your branch? This approach would mean we don't have to worry about changes to Stratigility (as I think it's fine to fetch all the middleware for a given `MiddlewarePipe` at once in this circumstance), and I think the feature has some cool benefits; I'm already thinking about how we could have the `ApplicationFactory` also do global routed-middleware, into which the matched middleware would be injected. :smile: